### PR TITLE
feat(cam_models, examples): linear sphere camera model.

### PR DESCRIPTION
**Pull request by yoraish**
Once upon a time, there was a magical realm where all spheres underwent a series of dimensionality reduction projections. They all became lines, at the end. The name of their world was called the "Linear Sphere World." This PR is in their memory.

More concretely, this pull request introduces a simple **LinearSphere** camera model. This camera model places a pixel on the image plane in a position that moves linearly with its spherical coordinate.

### CODE REVIEW

File: camera_models.py [hunk 1]
```diff
@@ -1129,4 +1130,171 @@ def get_rays_wrt_lidar_frame(self):
...
+        
+        # Pixel coordinates.
+        if normalized:
+            u = u / self.ss.W * 2 - 1
+            v = v / self.ss.H * 2 - 1
+
+        pixel_coor = torch.cat( (u, v), dim=-2 )
+
+        # Filter by FOV.
+        a = x2y2z_2_z_angle( x2, y2, z )
```
Comment (huyaoyu) [inline camera_models.py]: Filter by FOV may not be needed?

File: camera_models.py [hunk 2]
```diff
@@ -1129,4 +1130,171 @@ def get_rays_wrt_lidar_frame(self):
...
+        x = r_xy * torch.cos( a_x )
+        y = r_xy * torch.sin( a_x )
+        
+        # Create an array of 3D points at the unit sphere along the FOV.
+        xyz = torch.cat( (x, y, z), dim=0 )
+        
+        pixel_coor, mask = self.point_3d_2_pixel(xyz)
+        return self.out_wrap(pixel_coor)
+    
+    def __str__(self) -> str:
```
Comment (huyaoyu) [inline camera_models.py:1284]: This needs to be revised. The member variables are wrong.

I am not so familiar with the `__repr__` and `__str__` magic functions. May be we need to revise all of them. I have the impression that `__repr__` should try to generate a string that can be used to re-generate the underlying object as python codes. `__str__` is some pretty string that friendly to humans for reading and understanding the object.

File: camera_models.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 
 import copy
+import numpy as np
```
Comment (huyaoyu) [inline camera_models.py]: Is NumPy needed here?

**Review by huyaoyu — COMMENTED** [review]
Please see the comments. Thank you.

File: camera_models.py [hunk 4]
```diff
@@ -1129,4 +1130,171 @@ def get_rays_wrt_lidar_frame(self):
...
+        # The pixel coordinates.
+        # If any are FTensors, then convert to Tensor.
+        if isinstance(th, FTensor):
+            th = th.tensor()
+            ph = ph.tensor()
+
+
+        u = r * torch.cos(th) + self.ss.W / 2
+        v = r * torch.sin(th) + self.ss.H / 2
+
```
Comment (yoraish) [inline camera_models.py:1242]: @huyaoyu What do you think about this?
I added this piece, where FTensors are converted to torch Tensors, since FTensors do not allow element-wise multiplications.
Does it make sense to do it like this? It feels a bit hacky to me.

File: camera_models.py (hunk 3 above)
Comment (yoraish) [reply camera_models.py]: Nop! Took it out.

File: camera_models.py (hunk 1 above)
Comment (yoraish) [reply camera_models.py]: Why not?
What I am thinking is that without it points behind the camera might still be allowed to be projected onto the frame. I didn't think too deeply about this -- is there something I am missing?

File: camera_models.py (hunk 2 above)
Comment (yoraish) [reply camera_models.py:1284]: Oh whoops yes I forgot about this one. Thank you for catching it!
And generally, I mostly try to write a `__str__` for `print` calls. We could revise the `__repr__` if you'd like.

File: camera_models.py (hunk 1 above)
Comment (huyaoyu) [reply camera_models.py]: On Line 1231 of the current file
```python
r = ph * self.ss.W / self.fov_rad
```
This asures that the border of the image is at `self.fov_rad / 2 angle`. And Line 1244
```python
valid_mask = (u >= 0) & (u < self.ss.W) & (v >= 0) & (v < self.ss.H)
```
makes sure that anything beyond the image border (larger than the FoV) will be masked.

File: camera_models.py (hunk 4 above)
Comment (huyaoyu) [reply camera_models.py:1242]: This is in fact the correct way!

I force the user of ftensor to be absolutely explicit about what they want to do. That is, if two ftensors get multiplied, then it must mean a transform. If it is not a transform, `.tensor()` should not be used to clearly signal the developer her/himself and future developers that the multiplication breaks the state of being an ftensor. They need to handle the frame definitions by themselves then.

**Review by huyaoyu — COMMENTED** [review]
Make new comments.

File: camera_models.py (hunk 1 above)
Comment (yoraish) [reply camera_models.py]: Great! This is what I was missing. Will take it out.

File: camera_models.py (hunk 4 above)
Comment (yoraish) [reply camera_models.py:1242]: Awesome!

**Comment by yoraish** [comment]
Thank you for the helpful comments, Yaoyu!

All comments have been addressed, this branch will be merged and deleted.
